### PR TITLE
new feature: `docker -H ssh://host`

### PR DESCRIPTION
**Pull request by AkihiroSuda**
**- What I did**

Implemented `docker -H ssh://host`, which allows connecting to a Unix socket listening on a remote daemon via an OpenSSH connection.

```
[docker] ----- TCP -----> [sshd] -- Unix --> [dockerd]
```

Replaces https://github.com/moby/moby/pull/32161 (which used Go-native SSH rather than `os/exec` even on Unix, and hence didn't support `~/.ssh/config`)

Closes https://github.com/docker/docker/issues/31871

Motivations:

- No mess of setting up TLS properly (`ssh-keygen` vs a bunch of [openssl blah -blah -blah -blah ...](https://docs.docker.com/engine/security/https/#create-a-ca-server-and-client-keys-with-openssl), #6817)
- Single user account management for both regular SSH session and Docker session
- Single login monitoring for both

The feature is purely implemented on the client side.
No need to configure the daemon.

**- How I did it**

- on Unix: By `os/exec`-uting `ssh -L local.sock:/var/run/docker.sock`. So it should even work with `~/.ssh/config`.

- on Windows: using pageant + Go-native ssh (golang.org/x/crypto/ssh)

**- How to verify it**
 
* No need to configure the daemon. Even no need to replace the daemon built from this PR.  I confirmed this PR works well with vanilla 17.05.0-ce daemon.

* You just need to create a SSH user account that can access `/var/run/docker.sock`

* On Windows client, you need to register your key to `pageant.exe`. No need to install `ssh.exe`.

```console
$ go build ./contrib/client-test
$ ./client-test ssh://me@host
(JSON representation of `docker info` will be shown)
```

The argument can be also:
- `ssh://me@host:22/var/run/docker.sock`
- `ssh://host` (using `$USER` as the user name)
- ...

**- Description for the changelog**
Now docker client can connect to a Unix socket listening on a remote daemon via an OpenSSH connection

**- A picture of a cute animal (not mandatory but encouraged)**

penguins

TODO
- [ ] Vendor https://github.com/docker/go-connections/pull/39

Signed-off-by: Akihiro Suda <suda.akihiro@lab.ntt.co.jp>

### CODE REVIEW

**Comment by AkihiroSuda** [comment]
cc-ing those who has commented on #32161

@vdemeester @n4ss @diogomonica @thaJeztah @justincormack @friism @nathanleclaire @tonistiigi @jhowardmsft @arianitu @rca @stevvooe @chungers

File: opts/hosts.go [hunk 1]
```diff
@@ -77,6 +79,12 @@ func parseDockerDaemonHost(addr string) (string, error) {
 		return parseSimpleProtoAddr("npipe", addrParts[1], DefaultNamedPipe)
 	case "fd":
 		return addr, nil
+	case "ssh": // unix over ssh (can be specified for `docker -H` but not for `dockerd -H`)
+		usr, err := whoami()
```
Comment (tonistiigi) [inline opts/hosts.go]: Why not let `ssh(1)` handle this? This calls into `/etc/passwd` parsing on `macOS` that doesn't work for regular users.

Comment (AkihiroSuda) [inline opts/hosts.go]: ok, I'll update

**Comment by justincormack** [comment]
Given that Windows will soon have native openssh support, it seems weird to special case it now?

**Comment by AkihiroSuda** [comment]
@justincormack 
If we are going to use `socat` or `nc` instead of `-L local.sock:/var/run/docker.sock`, we can eliminate special case for Windows ( https://github.com/docker/go-connections/pull/39#issuecomment-307031239 )

**Comment by AkihiroSuda** [comment]
Method | No socket file leak | Support `~/.ssh/config` | No dependency on Docker product | Likely to work on default Docker installation
--- | --- | --- | --- | --- 
`golang.org/x/crypto/ssh` | :white_check_mark: | :x:  | :white_check_mark: | :white_check_mark: |
`golang.org/x/crypto/ssh`+implement `~/.ssh/config` parser | :white_check_mark: | :white_check_mark: (hard)  | :white_check_mark: | :white_check_mark: |
`os/exec("ssh -L local.sock:/var/run/docker.sock)` | :x: | :white_check_mark:  | :white_check_mark: | :white_check_mark: |
`os/exec("ssh socat-or-nc")` | :white_check_mark: | :white_check_mark:  | :white_check_mark: | :x: (`socat` and `nc` are rarelly installed) |
`os/exec("ssh docker sockproxy")` | :white_check_mark: | :white_check_mark:  | :x: | :white_check_mark: |
`os/exec("ssh dockerd sockproxy")` | :white_check_mark: | :white_check_mark:  | :x: (`dockerd` binary name might change on Moby Project) | :white_check_mark: |
`os/exec("ssh docker sockproxy \|\| ssh socat-or-nc")` | :white_check_mark: | :white_check_mark:  | :white_check_mark: | :white_check_mark: |
SSH subsystem + `os/exec("ssh -s somesubsystem")` | :white_check_mark: | :white_check_mark:  | :white_check_mark: | :x: (needs to modify `sshd_config`) |

design discussion held at https://github.com/docker/go-connections/pull/39

Comments are welcome at https://github.com/docker/go-connections/pull/39

**Comment by cpuguy83** [comment]
What do you think about implementing a "connection helpers" interface like we do for credentials.
This would allow a user to supply their own custom helpers that will just work natively with the docker CLI.
Helpers can be defined in `~/.docker/config.json` matching the protocol of the connection string to a config.
e.g.

```json
{
  "ssh": "~/.docker/ssh_connection_helper",
  "weird": "~/my_weird_connection_helper"
}
```

Where `-H ssh://1.2.3.4:22` would execute the `~/.docker/ssh_connection_helper` helper and `-H weird://1.2.3.4` would execute `~/my_weird_connection_helper`

**Comment by AkihiroSuda** [comment]
@cpuguy83 SGTM, I'll try to update this PR.
Do you come up with other possible plugins?

**Comment by AkihiroSuda** [comment]
Temporarily closing.

Further discussion should be at https://github.com/moby/moby/issues/31871

**Comment by AkihiroSuda** [comment]
Implemented @cpuguy83 's proposal: https://github.com/docker/cli/pull/889

**Pull request closed without merging**